app/handlers/sessions_create.rb

```diff
@@ -120,23 +120,36 @@ def handle_during_signup
   end
 
   def handle_while_logged_in
-    # Normally this happens when adding authentications on the profile screen.
+    # Attempt to login when already logged in
 
+    # Same user: don't do anything
     return :no_action if authentication_user == current_user
 
-    return :authentication_taken if authentication_user && authentication_user.is_activated?
+    # Check if they are trying to add a new authentication to their account
+    if request.env['omniauth.params'].try!(:[], 'add') == 'true'
+      # Add the new authentication
+      return :authentication_taken if authentication_user && authentication_user.is_activated?
 
-    return :same_provider \
-      if current_user.authentications.map(&:provider).include?(authentication.provider)
+      return :same_provider \
+        if current_user.authentications.map(&:provider).include?(authentication.provider)
 
-    return :new_signin_required if user_signin_is_too_old?
+      return :new_signin_required if user_signin_is_too_old?
 
-    return :email_already_in_use if ContactInfo.verified.where(value: @data.email)
-                                                        .where{user_id != my{current_user.id}}.any?
+      return :email_already_in_use if ContactInfo.verified.where(value: @data.email)
+                                                          .where{user_id != my{current_user.id}}
+                                                          .exists?
 
-    run(TransferAuthentications, authentication, current_user)
-    run(TransferOmniauthData, @data, current_user) if authentication.provider != 'identity'
-    return :authentication_added
+      run(TransferAuthentications, authentication, current_user)
+      run(TransferOmniauthData, @data, current_user) if authentication.provider != 'identity'
+      return :authentication_added
+    # Fallback to one of the other flows
+    elsif logging_in?
+      handle_during_login
+    elsif signing_up?
+      handle_during_signup
+    else
```

Comment:
Instead of having a fallback here, can we change the `handle` method to control which sub "handle" we call?

```ruby
def handle
  outputs.status = 
    if adding_authentication? # new method that is `signed_in? && see the add flag`
      handle_add_authentication
    elsif # actually is there a separate `handle_while_logged_in` case?
    elsif ...
...
```

Reply:
Also is the implementation of `signing_up?` still valid?  `pre_auth_state` can exist during login too?

Reply:
OK I will give combinining them a try.

I think signing_up? is fine because we check it last, so it only triggers if they are not logging in instead.